proxy/src/control/discovery.rs

```diff
@@ -499,7 +526,121 @@ where T: HttpService<RequestBody = BoxBody, ResponseBody = RecvBody>,
     }
 }
 
-// ===== impl RxError =====
+// ===== impl DstLabels ====
+
+impl DstLabels {
+
+    #[inline]
+    fn new<I, S>(labels: I) -> Option<Self>
+    where
+        I: IntoIterator<Item=(S, S)>,
+        S: fmt::Display,
+    {
+        let mut labels = labels.into_iter();
+
+        if let Some((k, v)) = labels.next() {
+            // format the first label pair without a leading comma, since we
+            // don't know where it is in the output labels at this point.
+            let mut s = format!("dst_{}=\"{}\"", k, v);
+
+            // format subsequent label pairs with leading commas, since
+            // we know that we already formatted the first label pair.
+            for (k, v) in labels {
+                write!(s, ",dst_{}=\"{}\"", k, v)
+                    .expect("writing to string should not fail");
+            }
+
+            Some(DstLabels(Arc::from(s)))
+        } else {
+            // the iterator is empty; return None.
+            None
+        }
+
+    }
+
+}
+
+impl fmt::Display for DstLabels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+
+}
+
+// ===== impl Labeled =====
+
+impl Labeled<SocketAddr> {
+
+    /// Construct a new labeled `SocketAddr `from a protobuf `WeightedAddr`.
+    fn from_pb(pb: WeightedAddr, set_labels: &Arc<HashMap<String, String>>)
+               -> Option<Self> {
+        let inner = pb.addr.and_then(pb_to_sock_addr)?;
+        let label_iter =
+            set_labels.as_ref()
+                .iter()
+                .chain(pb.metric_labels.iter());
+        let metric_labels = DstLabels::new(label_iter);
+        Some(Labeled { metric_labels, inner, })
+    }
+}
+
+
+impl<T> Labeled<T> {
+    /// Wrap `inner` with no `metric_labels`.
+    pub fn none(inner: T) -> Self {
+        Self { metric_labels: None, inner }
+    }
+
+    /// Construct a new `Labeled<U>` for `inner` with the same labels as `self`.
+    fn label<U>(&self, inner: U) -> Labeled<U> {
+        Labeled {
+            metric_labels: self.metric_labels.as_ref().cloned(),
+            inner,
+        }
+    }
+}
+
+impl<T> ops::Deref for Labeled<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        &self.inner
+    }
+}
+
+// By implementing `Borrow<SocketAddr>` for `Labeled<SocketAddr>`, we
+// can now treat a `SocketAddr` as a "borrowed form" of `Labeled<SocketAddr>`.
+// While this isn't technically what it is, that allows us to use an
+// un-labeled `SocketAddr` as a lookup key for `HashMap`s or `HashSet`s of
+// `Labeled<SocketAddr>`.
+impl<T> Borrow<T> for Labeled<T> {
+    fn borrow(&self) -> &T {
+        &self.inner
+    }
+}
+
+impl<T, A> Service for Labeled<T>
+where
+    T: Service<Request=http::Request<A>>,
+{
+    type Request = T::Request;
+    type Response= T::Response;
+    type Error = T::Error;
+    type Future = T::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
```

Comment:
It seems wrong to me that we'd increase the call stack depth (or rely heavily on inlining and COMDAT folding) just to get automated labeling of services. Again, this suggests to me that `Labeled` should be a trait implemented by labeled types of objects, and not a wrapper of such types.

Reply:
This is basically the strategy we expect for tower... that small middlewares can be used to keep orthogonal functionality separate.

How concerned do we really need to be about the callstack right now?  I have a strong preference for prioritizing readability, unless we are confident that the performance tradeoffs are substantial.

Reply:
I can definitely make this `#[inline]`

Reply:
Up to y'all.

The important thing to me is that we get rid of the `Deref` and `Borrow` implementations to avoid the dangerous implicit conversions.

Reply:
I would not prematurely add `inline` annotations.